server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+            } else {
+                logger.info("operator settings directory [{}] not found, will watch for its creation...", settingsDir);
+                enableSettingsWatcher(environment.configFile());
+            }
+        } catch (Exception e) {
+            if (watchService != null) {
+                try {
+                    this.watchService.close();
+                } catch (Exception ignore) {} finally {
+                    this.watchService = null;
+                }
+            }
+
+            throw new IllegalStateException("unable to launch a new watch service", e);
+        }
+
+        this.watcherThreadLatch = new CountDownLatch(1);
+
+        new Thread(() -> {
+            try {
+                logger.info("file settings service up and running [tid={}]", Thread.currentThread().getId());
+
+                WatchKey key;
+                while ((key = watchService.take()) != null) {
+                    /**
+                     * Reading and interpreting watch service events can vary from platform to platform. E.g:
+                     * MacOS symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_MODIFY:operator
+                     *     ENTRY_CREATE:settings.json
+                     *     ENTRY_MODIFY:settings.json
+                     * Linux in Docker symlink delete and set (rm -rf operator && ln -s <path to>/file_settings/ operator):
+                     *     ENTRY_CREATE:operator
+                     * After we get an indication that something has changed, we check the timestamp, file id,
+                     * real path of our desired file.
+                     */
+                    if (Files.exists(settingsDir)) {
+                        try {
+                            Path path = operatorSettingsFile();
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            enableSettingsWatcher(settingsDir);
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("unable to watch or read operator settings file", e);
+                        }
+                    } else {
+                        key.pollEvents();
+                        key.reset();
+                    }
+                }
+            } catch (Exception e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("encountered exception watching", e);
+                }
+                logger.info("shutting down watcher thread");
```

Comment:
This would only happen when an exception occurs, so is this useful as an info message? Why not log the exception at error level, stating the watcher thread is shutting down? Isn't this always an exception case (not how the thread would be shutdown)?

Reply:
I didn't want to pollute the logs with an exception trace when the watching is interrupted on the take() method call. It's how we interrupt the watcher when we stop watching, e.g. we are no longer the master node. So what I'll do here is log info when ClosedWatchServiceException is thrown and log error on any other exception.

Reply:
We could have a separate catch for InterruptedException so that any unknown errors are logged at the appropriate level?